### PR TITLE
Added tests for Book and Categories

### PR DESCRIPTION
**Pull request by IhorTrokhymchuk**

### CODE REVIEW

**Comment by IhorTrokhymchuk** [comment]
Screenshot 2024-02-26 at 17 26 01
Screenshot 2024-02-26 at 17 27 41

File: src/test/java/com/example/store/controller/CategoryControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    void setUp(@Autowired DataSource dataSource) throws SQLException {
+        teardown(dataSource);
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            ScriptUtils.executeSqlScript(connection,
+                    new ClassPathResource(
+                            "database/category/add-categories-to-categories-table.sql"));
+            ScriptUtils.executeSqlScript(connection,
+                    new ClassPathResource(
+                            "database/books/add-books-with-any-categories-to-book-table.sql"));
```
Comment (okuzan) [inline src/test/java/com/example/store/controller/CategoryControllerTest.java]: better extract to constants

File: src/test/resources/database/delete-all-data.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+delete from books_categories;
+delete from books;
+delete from categories;
```
Comment (okuzan) [inline src/test/resources/database/delete-all-data.sql]:
```suggestion
delete from categories;

```

File: src/test/resources/application.properties [hunk 3]
```diff
@@ -1,5 +1,6 @@
...
-spring.datasource.driverClassName=org.h2.Driver
-spring.datasource.username=sa
-spring.datasource.password=password
-spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
+spring.datasource.url=jdbc:tc:mysql:8:///test
+spring.datasource.username=test
+spring.datasource.password=test
+jwt.secret=mysecretkeyfortests213123123
+jwt.expiration=30012300123
+
```
Comment (okuzan) [inline src/test/resources/application.properties]:
```suggestion
```

File: src/test/java/com/example/store/service/CategoryServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+
+    @Test
+    @DisplayName("verify findAll method to give empty data")
+    void findAll_IfDataIsNotExist_ShouldThrowException() {
+        Pageable pageable = PageRequest.of(0,10);
+        Page<Category> page = new PageImpl<>(List.of());
+
+        when(categoryRepository.findAll(any(Pageable.class))).thenReturn(page);
+
+        assertThrows(EntityNotFoundException.class, () -> categoryService.findAll(pageable));
```
Comment (okuzan) [inline src/test/java/com/example/store/service/CategoryServiceTest.java]: it's normal for queries like findAll to not return anything, we shouldn't throw exception in this case

File: src/test/java/com/example/store/service/CategoryServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        assertEquals(expectedDto, findCategory);
+
+        verify(categoryRepository).findById(testCategoryId);
+        verify(categoryMapper).toDto(category);
+    }
+
+    @Test
+    @DisplayName("verify findById method to find by non valid id")
+    void findById_FindCategoryByNonValidId_ShouldThrowException() {
+        Long testCategoryId = Long.MAX_VALUE;
```
Comment (okuzan) [inline src/test/java/com/example/store/service/CategoryServiceTest.java]: a simple negative value would be more obvious, Long.MAX_VALUE technically could be a valid id

File: src/test/java/com/example/store/service/CategoryServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+        assertThrows(EntityNotFoundException.class, () -> categoryService.findAll(pageable));
+
+        verify(categoryRepository).findAll(any(Pageable.class));
+        verify(categoryMapper, times(page.getContent().size())).toDto(any(Category.class));
+        verifyNoMoreInteractions(categoryRepository, categoryMapper);
+    }
+
+    @Test
+    @DisplayName("verify getBooksByCategory method to give valid data")
+    void getBooksByCategory_GetExistData_ShouldReturnListBookDto() {
```
Comment (okuzan) [inline src/test/java/com/example/store/service/CategoryServiceTest.java]: use `existent` `nonexistent` for more clarity

File: src/test/java/com/example/store/service/CategoryServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        Pageable pageable = PageRequest.of(0,10);
+        Page<Category> page = new PageImpl<>(List.of(category1, category2));
+        when(categoryRepository.findAll(any(Pageable.class))).thenReturn(page);
+
+        when(categoryMapper.toDto(any(Category.class))).thenAnswer(
+                invocation -> {
+                    Category category = invocation.getArgument(0);
+                    CategoryDto categoryDto = new CategoryDto();
+                    categoryDto.setName(category.getName());
+                    categoryDto.setDescription(category.getDescription());
```
Comment (okuzan) [inline src/test/java/com/example/store/service/CategoryServiceTest.java:177]: looks like a repeated logic that could be extract to a private method

File: src/test/java/com/example/store/service/BookServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+                    bookDto.setAuthor(book.getAuthor());
+                    bookDto.setIsbn(book.getIsbn());
+                    bookDto.setPrice(book.getPrice());
+                    bookDto.setCategoryIds(book.getCategories().stream()
+                            .map(Category::getId)
+                            .collect(Collectors.toSet()));
+                    bookDto.setDescription(book.getDescription());
+                    bookDto.setCoverImage(book.getCoverImage());
+                    return bookDto;
+                }
```
Comment (okuzan) [inline src/test/java/com/example/store/service/BookServiceTest.java]: it's huge, extract to a private method

File: src/test/java/com/example/store/controller/CategoryControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        Long categoryIdToDelete = 1L;
+
+        MvcResult result = mockMvc.perform(delete("/categories/{id}", categoryIdToDelete)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNoContent())
+                .andReturn();
+    }
+
+    @Test
+    @DisplayName("Delete category by id with out authorities")
```
Comment (okuzan) [inline src/test/java/com/example/store/controller/CategoryControllerTest.java]: some of your display names start from lowercase, some from uppercase

File: src/test/java/com/example/store/controller/BookControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+        expectedDto.setIsbn(testIsbn);
+        expectedDto.setCategoryIds(testCategoryIds);
+        expectedDto.setPrice(testPrice);
+
+        CreateBookRequestDto requestDto = new CreateBookRequestDto();
+        requestDto.setTitle(testTitle);
+        requestDto.setAuthor(testAuthor);
+        requestDto.setIsbn(testIsbn);
+        requestDto.setCategoryIds(testCategoryIds);
+        requestDto.setPrice(testPrice);
```
Comment (okuzan) [inline src/test/java/com/example/store/controller/BookControllerTest.java]: test data setup takes a lot of lines, encapsulate it in a private method everywhere

File: src/test/java/com/example/store/controller/BookControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+import org.springframework.http.MediaType;
+import org.springframework.jdbc.datasource.init.ScriptUtils;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class BookControllerTest {
```
Comment (okuzan) [inline src/test/java/com/example/store/controller/BookControllerTest.java]: include Integration in name for clarity

**Review by okuzan — CHANGES_REQUESTED** [review]
GJ overall

**Review by okuzan — APPROVED** [review]
LGTM

**Pull request merged**